docs/proposals/job.md

```diff
@@ -0,0 +1,135 @@
+# Job Controller
+
+## Abstract
+A proposal for implementing a new controller - Job controller, which will be responsible
+for managing pod(s) that require to run on demand.
+
+Several existing issues and PRs were already created regarding that particular subject:
+* Job Controller [#1624](https://github.com/GoogleCloudPlatform/kubernetes/issues/1624)
+* New Job resource [#7380](https://github.com/GoogleCloudPlatform/kubernetes/pull/7380)
+
+
+## Use Cases
+1. Be able to start of one or several pods tracked as a single entity.
+1. Be able to implement heavy-computation model, eg. MapReduce.
+1. Be able to get the actual job status.
+1. Be able to limit the execution time for a job.
+
+
+## Motivation
+Jobs are needed for executing multi-pod computation "to completion", a good example
+here would be the ability to implement a MapReduce or Hadoop style workload.
+Additionally this new controller should take over pod management logic we currently
+have in certain OpenShift controllers, namely build controller.
+
+
+## Implementation
+Job controller, is similar to replication controller in that they manage pods
+representing a `Job`.  This implies they will follow the same controller framework
+that replication controllers already defined.  The biggest difference between
+`Job` and `ReplicationController` objects is the `RestartPolicy` which is required
+to always take value of `RestartPolicyNever` or `RestartOnFailure`.  The main
+responsibility of a job controller will be keeping that the number of pods will
+match the number specified in the job specification in such way that the job
+should finish its calculation.
+
+
+The new `Job` object will have the following content:
+
+```go
+// Job represents the configuration of a single job.
+type Job struct {
+    TypeMeta
+    ObjectMeta
+
+    // Spec is a structure defining the expected behavior of a job.
+    Spec JobSpec
+
+    // Status is a structure describing current status of a job.
+    Status JobStatus
+}
+
+// JobList is a collection of jobs.
+type JobList struct {
+    TypeMeta
+    ListMeta
+
+    Items []Job
+}
+```
+
+`JobSpec` structure is defined to contain all the information how the actual job execution
+will look like.
+
+```go
+// JobSpec describes how the job execution will look like.
+type JobSpec struct {
+
+    // PodCount specifies the desired number of pods the job should be run with.
+    PodCount int
```

Comment:
At this point - maybe let's move this proposal to Kube and record brian's
comments.  I think the initial cleanup of the proposal was enough but this
should be in the wider audience (now that folks have time to review
upstream).

On Wed, Jul 22, 2015 at 1:44 PM, Brian Grant <notifications@github.com>
wrote:

> In docs/proposals/job.md
> <https://github.com/openshift/origin/pull/3693#discussion_r35241384>:
>
> > +    TypeMeta
> > +    ListMeta
> > +
> > +    Items []Job
> > +}
> > +```
> > +
> > +`JobSpec` structure is defined to contain all the information how the actual job execution
> > +will look like.
> > +
> > +```go
> > +// JobSpec describes how the job execution will look like.
> > +type JobSpec struct {
> > +
> > +    // PodCount specifies the desired number of pods the job should be run with.
> > +    PodCount int
>
> TaskCount would be more descriptive in the case that it is intended to
> represent the desired number of completed tasks.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3693/files#r35241384>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift